feat: auto-connect on startup

re-vc opened the pull request:
hi, this is a simple "auto join to wynncraft on startup" feature.
configurable - off by default.
tested and working on forge, fabric and quill - jdk 17.0.7
i considered splitting it off into a separate class but let me know what you think.

re-vc committed:
feat: auto-connect on startup

re-vc committed:
fix: issue with autoConnect reconnection after disconnecting from server

re-vc committed:
Merge branch 'main' into autoconnect

ryanzhoudev commented on common/src/main/java/com/wynntils/features/ui/WynncraftButtonFeature.java:
```diff
@@ -38,15 +38,20 @@
 public class WynncraftButtonFeature extends Feature {
     private static final String GAME_SERVER = "play.wynncraft.com";
     private static final String LOBBY_SERVER = "lobby.wynncraft.com";
+    private static boolean initialBoot = false;
```
Does this boolean mark that it is currently the initial boot, or that the initial boot has passed? Change the name to be a bit more clear please

re-vc replied:
its called "initBootPassed" now - and it marks if the TitleScreen was "loaded?" at least once.
feel free to suggest a different name for the variable

ryanzhoudev commented on common/src/main/java/com/wynntils/features/ui/WynncraftButtonFeature.java:
```diff
@@ -66,6 +71,22 @@ private void addWynncraftButton(TitleScreen titleScreen) {
 
+    /* connects to the server on "startup" - off by default */
+    private void autoConnect() {
+        if (!initialBoot && autoConnect.get()) {
```
Can this config check be moved to where autoConnect() is called, so it forms a natural flow? Also, we don't need to add the wynncraft button if we are immediately joining the server, so maybe put that in some check as well.

kristofbolyai replied:
Moving the check is enough, no need to complicate it further

ryanzhoudev commented on common/src/main/java/com/wynntils/features/ui/WynncraftButtonFeature.java:
```diff
@@ -66,6 +71,22 @@ private void addWynncraftButton(TitleScreen titleScreen) {
+        if (!initialBoot && autoConnect.get()) {
+            ServerData wynncraftServer =
+                    new ServerData("Wynncraft", connectToLobby.get() ? LOBBY_SERVER : GAME_SERVER, false);
+            wynncraftServer.setResourcePackStatus(ServerData.ServerPackStatus.ENABLED);
```
Has this been checked in both cases where the mod loads the resource pack, and where the mod does not?

re-vc replied:
if you refer to whether or not it causes any unexpected behavior ive tested it with and without mod loaded resource pack and it seems to be fine.

ryanzhoudev commented on common/src/main/resources/assets/wynntils/lang/zh_cn.json:
```diff
@@ -686,6 +686,8 @@
   "feature.wynntils.worldWaypointDistance.textShadow.name": "文本阴影",
   "feature.wynntils.worldWaypointDistance.topBoundingDistance.description": "离屏幕时，航路点指示器最近对屏幕的顶部距离是多少？",
   "feature.wynntils.worldWaypointDistance.topBoundingDistance.name": "顶部提示器距离",
+  "feature.wynntils.wynncraftButton.autoConnect.description": "translate: Should you auto-connect to Wynncraft on startup?",
```
Also, I am not sure we need these. I believe most features are merged with only an entry for en_us.json.

kristofbolyai replied:
Remove these, only keep en-us

kristofbolyai commented on common/src/main/java/com/wynntils/features/ui/WynncraftButtonFeature.java:
```diff
@@ -38,15 +38,20 @@
 public class WynncraftButtonFeature extends Feature {
     private static final String GAME_SERVER = "play.wynncraft.com";
     private static final String LOBBY_SERVER = "lobby.wynncraft.com";
+    private static boolean initialBoot = false;
```
```suggestion
    private boolean initialBoot = false;
```

Feature variables musn't be static

kristofbolyai commented on common/src/main/java/com/wynntils/features/ui/WynncraftButtonFeature.java:
```diff
@@ -66,6 +71,22 @@ private void addWynncraftButton(TitleScreen titleScreen) {
+            wynncraftServer.setResourcePackStatus(ServerData.ServerPackStatus.ENABLED);
+
+            Minecraft minecraft = Minecraft.getInstance();
+            Screen backScreen = minecraft.screen;
```
McUtils.mc().screen

kristofbolyai commented on common/src/main/java/com/wynntils/features/ui/WynncraftButtonFeature.java:
```diff
@@ -66,6 +71,22 @@ private void addWynncraftButton(TitleScreen titleScreen) {
+            ConnectScreen.startConnecting(
+                    backScreen, minecraft, ServerAddress.parseString(wynncraftServer.ip), wynncraftServer);
+        }
+        initialBoot = true;
```
Just set this before you start connecting so this only happens once?

re-vc commented:
Thank you for the feedback. I appreciate your input, and I will do my best to incorporate the suggested changes this week.

re-vc committed:
fix: clearer variable naming, moved check, optimized.

re-vc committed:
Merge remote-tracking branch 'origin/autoconnect' into autoconnect

ryanzhoudev commented on common/src/main/java/com/wynntils/features/ui/WynncraftButtonFeature.java:
```diff
@@ -53,7 +58,12 @@ public void onTitleScreenInit(TitleScreenInitEvent.Post e) {
+        } else {
+            addWynncraftButton(titleScreen);
+        }
+        initBootPassed = true;
```
What is the purpose of this? Doesn't the autoConnect() function already set this to be true?

re-vc replied:
Having inside of the onTitleScreenInit() is crucial - it ensures that if the user goes back to the Title-Screen by any means - they will not get autoConnected to the server.

technically we can take it out in the autoConnect() method.
kristofbolyai said "Just set this before you start connecting so this only happens once?" 
i think it may or may not ensure to not try to connect multiple times at once or in rapid succession.

ryanzhoudev replied:
I see. Then, we could put it before that if statement and remove the one inside autoConnect()?

re-vc replied:
I think the best thing would be, removing it from autoConnect() and leaving it in the onTitleScreenInit() but after the if statement - if put above. it would never trigger autoConnect().

Another option would be splitting the if statement which consists of the config-check && initBootPassed and move the initBootPassed check to autoConnect()

in either case the variable will be updated to true each time if on titleScreen

ryanzhoudev commented on common/src/main/java/com/wynntils/features/ui/WynncraftButtonFeature.java:
```diff
@@ -66,6 +76,23 @@ private void addWynncraftButton(TitleScreen titleScreen) {
     }
 
+    /** connects to the server on "startup".
+     *  requires the WynncraftButtonFeature to be enabled.
```
```suggestion
     *  Requires the WynncraftButtonFeature to be enabled.
```

ryanzhoudev commented on common/src/main/java/com/wynntils/features/ui/WynncraftButtonFeature.java:
```diff
@@ -66,6 +76,23 @@ private void addWynncraftButton(TitleScreen titleScreen) {
         titleScreen.addRenderableWidget(wynncraftButton);
     }
 
+    /** connects to the server on "startup".
```
```suggestion
    /** Connects to the server on "startup".
```

re-vc committed:
Update common/src/main/java/com/wynntils/features/ui/WynncraftButtonFeature.java

Co-authored-by: Ryan <57310593+DonkeyBlaster@users.noreply.github.com>

re-vc committed:
Update common/src/main/java/com/wynntils/features/ui/WynncraftButtonFeature.java

Co-authored-by: Ryan <57310593+DonkeyBlaster@users.noreply.github.com>

re-vc committed:
fix: double initBootPassed = true

ryanzhoudev reviewed (approved):
Looks fine to me, though I wouldn't be opposed to inverting initBootPassed to isInitBoot and so on if you wanted to.

kristofbolyai commented on common/src/main/java/com/wynntils/features/ui/WynncraftButtonFeature.java:
```diff
@@ -66,6 +76,21 @@ private void addWynncraftButton(TitleScreen titleScreen) {
 
+    /** Connects to the server on "startup".
+     *  Requires the WynncraftButtonFeature to be enabled.
+     */
```
There is no reason to have this comment. More is not always better, and that is the case here.  All the info you describe here is true but can be collected from the feature's class itself.

re-vc replied:
removed.
it was a wip comment anyway but the second part the "Requires the WynncraftButtonFeature to be enabled." 
Ive actually spent easily an hour or two to find out.

kristofbolyai commented on common/src/main/java/com/wynntils/features/ui/WynncraftButtonFeature.java:
```diff
@@ -66,6 +76,21 @@ private void addWynncraftButton(TitleScreen titleScreen) {
+     *  Requires the WynncraftButtonFeature to be enabled.
+     */
+    private void autoConnect() {
+        ServerData wynncraftServer =
+                new ServerData("Wynncraft", connectToLobby.get() ? LOBBY_SERVER : GAME_SERVER, false);
+        wynncraftServer.setResourcePackStatus(ServerData.ServerPackStatus.ENABLED);
```
Since this is straight up copied from above, can you extract this into a method and get the `ServerData` instance from there?

re-vc committed:
fix: moved fetching ServerData into getWynncraftServer().

magicus commented on common/src/main/java/com/wynntils/features/ui/WynncraftButtonFeature.java:
```diff
@@ -53,19 +58,40 @@ public void onTitleScreenInit(TitleScreenInitEvent.Post e) {
+        if (!initBootPassed && autoConnect.get()) {
+            autoConnect();
+        } else {
+            addWynncraftButton(titleScreen);
```
Why is this an else? I presume that you will want the button added even if you autoconnect, so it is there after you log out from Wynncraft.

re-vc replied:
The "addWynncraftButton()" method is only necessary for rendering the button on the initial start of the game. If you want to autoconnect during the initial start, you don't need to include it. The button is added through the "onTitleScreenInit(TitleScreenInitEvent.Post e)" event, which triggers after logging out or returning to the titlescreen from any other menu.

magicus commented on common/src/main/java/com/wynntils/features/ui/WynncraftButtonFeature.java:
```diff
@@ -53,19 +58,40 @@ public void onTitleScreenInit(TitleScreenInitEvent.Post e) {
+    private void autoConnect() {
+        ServerData wynncraftServer = getWynncraftServer();
+
+        ConnectScreen.startConnecting(
```
This is shared with the `onPress` method in the button class. Please extract it to a common private method.

And after doing that, `autoConnect` becomes only two lines, and is not needed (can be inlined in place).

re-vc committed:
fix: separate connectToServer() used in autoConnect and onPress.
BREAKING CHANGE: WynncraftButton Constructor is now using a local Variable backScreen.

kristofbolyai commented on common/src/main/java/com/wynntils/features/ui/WynncraftButtonFeature.java:
```diff
@@ -38,10 +39,14 @@
 public class WynncraftButtonFeature extends Feature {
     private static final String GAME_SERVER = "play.wynncraft.com";
     private static final String LOBBY_SERVER = "lobby.wynncraft.com";
+    private boolean hasInitBootPased = false;
```
This variable has a spelling mistake and it is impossible to figure out what it means (without reading usages of the variable).

Please rename this to `firstTitleScreenInit` or something with the same meaning

re-vc replied:
got it, using firstTitleScreenInit now, its a good name.
also sorry i just missed that spelling error somehow.

kristofbolyai commented on common/src/main/resources/assets/wynntils/lang/en_us.json:
```diff
@@ -868,6 +868,8 @@
   "feature.wynntils.wybelSound.hideText.description": "Should we hide the chat text if we play a sound?",
   "feature.wynntils.wybelSound.hideText.name": "Hide Chat Text",
   "feature.wynntils.wybelSound.name": "Wybel Sounds",
+  "feature.wynntils.wynncraftButton.autoConnect.description": "Should you auto-connect to Wynncraft on startup?",
```
```suggestion
  "feature.wynntils.wynncraftButton.autoConnect.description": "Should the mod auto-connect to Wynncraft on startup?",
```

kristofbolyai commented on common/src/main/java/com/wynntils/features/ui/WynncraftButtonFeature.java:
```diff
@@ -53,29 +58,44 @@ public void onTitleScreenInit(TitleScreenInitEvent.Post e) {
+        } else {
+            addWynncraftButton(titleScreen);
+        }
+        hasInitBootPased = true;
```
Refactor this into:

```
if firstboot and autoconnect:
then set firstboot to false;
         connecttoserver;
         return

addwynncraftbutton
```

re-vc replied:
i see now i get what DonkeyBlaster meant by inverting the firstTitleScreenInit!

re-vc committed:
Update common/src/main/resources/assets/wynntils/lang/en_us.json

Co-authored-by: Kristof Kovacs <49001742+kristofbolyai@users.noreply.github.com>

re-vc committed:
fix: Renamed to firstTitleScreenInit and inverted the check.

re-vc committed:
Merge remote-tracking branch 'origin/autoconnect' into autoconnect

kristofbolyai commented:
Thanks for working with us on this.
